src/components/molecules/denomination-grid/index.tsx

```diff
@@ -0,0 +1,83 @@
+import React, { useEffect, useMemo, useRef, useState } from "react"
+import { useObserveWidth } from "../../../hooks/use-observe-width"
+import DenominationBadge from "../../atoms/denomination-badge"
+
+export type GiftCardVariant = {
+  prices: {
+    currency_code: string
+    amount: number
+  }[]
+}
+
+type DenominationGridProps = {
+  variants: GiftCardVariant[]
+  defaultCurrency: string
+}
+
+const DenominationGrid: React.FC<DenominationGridProps> = ({
+  variants,
+  defaultCurrency,
+}) => {
+  const containerRef = useRef(null)
+
+  const width = useObserveWidth(containerRef)
+
+  const [columns, setColumns] = useState(7)
+
+  // We estimate that a Badge is rouglhy 70px and use this to calulate how many we can fit in a row
+  // It is a bit of a "magic" number, but it works for now.
+  useEffect(() => {
+    if (width) {
+      setColumns(Math.floor(width / 70) - 1)
+    }
+  }, [width])
+
+  const defaultDenominations = useMemo(() => {
+    return (
+      variants.map(variant => {
+        const price = variant.prices.find(
+          price => price.currency_code === defaultCurrency
+        )
+        return { amount: price?.amount, currencyCode: defaultCurrency }
+      }) ?? []
+    )
+  }, [variants, defaultCurrency])
+
+  const [denominations, setDenominations] = useState<
+    { amount: number; currencyCode: string }[] | null
+  >([])
+  const [remainder, setRemainder] = useState(0)
+
+  useEffect(() => {
+    if (defaultDenominations.length) {
+      setDenominations(defaultDenominations.slice(0, columns))
+      setRemainder(defaultDenominations.length - columns)
+    }
```

Comment:
nit: I believe the states here (+columns) could be derived because everything other piece of state is derived from the `width` we get from the `useObserveWidth`. Meaning this would be enough, no `useEffect`s needed:
```jsx
const columns = Math.max(Math.floor(width / 70) - 1, 1)
const denominations = defaultDenominations.slice(0, columns)
const remainder = defaultDenominations.length - columns
```

Reply:
even smaller nit: additionally, you could extract everything related to the denominations in a `useDenominationClamp` and just use it in the `DenominationGrid`. To what extent you'd want it to be reusable is up to you, but simply extracting into a hook without much re-usability would be good enough to clean up the `DenominationGrid` I think